Adding coverage report for server-side tests using istanbul

lirantal opened the pull request:
Code coverage is now built-in to our project and grunt tasks!

![meanjs-code-coverage-2](https://cloud.githubusercontent.com/assets/316371/9290945/2b2d8c10-43ae-11e5-8319-03ddb3762350.PNG)

and further results:

![meanjs-code-coverage-1](https://cloud.githubusercontent.com/assets/316371/9290946/2dc888a8-43ae-11e5-90c7-5b8fa57c5a80.PNG)

To run the code coverage simply execute:
```bash
grunt coverage
```

Addresses:
* https://github.com/meanjs/mean/pull/569
* https://github.com/meanjs/mean/pull/760

lirantal commented:
@ilanbiala please go ahead and commit your coveralls to this PR

mleanos commented:
LGTM. Really slick! I really like the initializing status :+1:

lirantal commented:
lol cool

ilanbiala commented:
@lirantal I won't be able to test anything until the weekend, so until then if you want to take a crack at it with Coveralls' guide, go for it.

lirantal commented:
@ilanbiala so let's merge this PR and we can later create a new one to extend the local coverage we're running with an integration to coveralls.

lirantal commented:
@codydaig @mleanos do you guys want to review this PR? I thought maybe you have some experience with code coverage tools

lirantal commented:
@sielay @abhijt-zanak maybe you also have something to add?

ilanbiala commented:
@lirantal I'd rather add it all at once. I don't think people are really gonna benefit from just having coverage in the terminal compared to having Coveralls running for them.

lirantal commented:
@ilanbiala WDYT?
we must merge this first to `master` for coveralls.io to actually pick it up and show information about code coverage

codydaig commented on README.md:
```diff
@@ -3,6 +3,7 @@
 [![Gitter](https://badges.gitter.im/Join Chat.svg)](https://gitter.im/meanjs/mean?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge)
 [![Build Status](https://travis-ci.org/meanjs/mean.svg?branch=master)](https://travis-ci.org/meanjs/mean)
 [![Dependencies Status](https://david-dm.org/meanjs/mean.svg)](https://david-dm.org/meanjs/mean)
+[![Coverage Status](https://coveralls.io/repos/meanjs/mean/badge.svg?branch=master&service=github)](https://coveralls.io/github/meanjs/mean?branch=coveralls-code-coverage)
```
@lirantal The .svg is displaying from the master branch but the link is going to the coveralls-code-coverage branch.

lirantal replied:
thanks! that's probably a leftover from taking README update off of ilan's branch.

lirantal replied:
there's also a missing `service=github` in the actual link?

codydaig replied:
I'm not entirely sure about that one, I just noticed the odd ball branch.

lirantal replied:
nm, looked at their website, it's not required

mleanos commented:
@lirantal LGTM... tested and verified the coverage outputs the logs correctly. Everything looks good.

And *STILL* like the Initializing status :)

lirantal commented:
lol @mleanos  :)
Awesome, thanks!

ilanbiala commented on gruntfile.js:
```diff
@@ -176,6 +176,22 @@ module.exports = function (grunt) {
+          coverage: true,
+          require: 'test.js',
+          coverageFolder: 'coverage',
+          reportFormats: ['cobertura','lcovonly'],
```
What's cobertura?

ilanbiala commented on gruntfile.js:
```diff
@@ -176,6 +176,22 @@ module.exports = function (grunt) {
+          coverageFolder: 'coverage',
+          reportFormats: ['cobertura','lcovonly'],
+          check: {
+            lines: 40,
```
Do we want these limits?

ilanbiala commented on package.json:
```diff
@@ -30,6 +30,7 @@
     "connect-mongo": "~0.8.1",
     "consolidate": "~0.13.1",
     "cookie-parser": "^1.3.2",
+    "coveralls": "^2.11.4",
```
should be a devDep no?

ilanbiala commented on test.js:
```diff
@@ -0,0 +1,13 @@
+var app = require(path.resolve('./config/lib/app'));
+
+app.init(function (app, db, config) {
+	console.log('Initialized test automation');
```
indent

ilanbiala commented on test.js:
```diff
@@ -0,0 +1,13 @@
+/**
+ * Module dependencies.
+ */
+var app, db, config;
```
Why are db and config set up here but they don't seem to be used anywhere.

lirantal commented:
@ilanbiala per your questions:
1. Cobertura is a code coverage tool which is commonly used in Java project. the grunt-mocha-istanbul plugin supports exports to the cobertura reporter format to integrate with hudson/jenkins. We use in our enterprise both Coberutra and Jenkins so I figure others will find this useful too without requiring extra tweaks.
2. About the check limits - for now we should keep that threshold because we don't have much more than that covered, so until we have more code coverage let's keep it on a 'sane' threshold. (I don't see a reason to keep a 90 threshold to show the project as not reaching it's goal and keeping that for who knows how long...
3. Moved coveralls to devDep
4. Corrected indentations
5. Fixed

codydaig commented:
LGTM

ilanbiala commented:
LGTM, @lirantal do we already see some coverage reports for this PR or not until we merge?

lirantal commented:
After we'll merge (I'll do it later), and then we can also tweak and take care of areas for more test as we wish.

ilanbiala commented:
Sounds good.

ilanbiala commented:
@lirantal are we merging this or is there something left to add?

lirantal commented:
Merging! :)

ilanbiala commented:
@lirantal Coveralls still doesn't show anything https://coveralls.io/github/meanjs/mean when I log in and see the repo status. Is there something else we have to add?

lirantal commented:
I'll check this

lirantal commented:
@ilanbiala ok so I put a specific grunt task called coverage and we aren't calling it as part of the travis ci tests today, and I don't think we should. Instead, we should add another task to the Travis CI to run the coverage. I'll work it out and let you know.

ilanbiala commented:
@lirantal :+1:
